src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -595,81 +601,112 @@ private bool IsBaseOrIdentity(ITypeSymbol source, ITypeSymbol dest)
+
+                    if (closeBraceIndex <= -1) // unclosed '{'
                     {
-                        // Format item syntax : { index[,alignment][ :formatString] }.
-                        int formatDelimiterIndex = FindIndexOfAny(message, _formatDelimiters, openBraceIndex, closeBraceIndex);
+                        success = false;
+                        break;
+                    }
+
+                    // Format item syntax : { index[,alignment][ :formatString] }.
+                    int formatDelimiterIndex = FindIndexOfAny(message, _formatDelimiters, openBraceIndex, closeBraceIndex);
+                    string templateName = message.Substring(openBraceIndex + 1, formatDelimiterIndex - openBraceIndex - 1);
 
-                        string templateName = message.Substring(openBraceIndex + 1, formatDelimiterIndex - openBraceIndex - 1);
-                        templateMap[templateName] = templateName;
-                        templateList.Add(templateName);
-                        scanIndex = closeBraceIndex + 1;
+                    if (string.IsNullOrWhiteSpace(templateName)) // braces with no named argument, such as {} and { }
+                    {
+                        success = false;
+                        break;
                     }
+
+                    templateMap[templateName] = templateName;
+                    templateList.Add(templateName);
+
+                    scanIndex = closeBraceIndex + 1;
                 }
+
+                return success;
             }
 
-            private static int FindBraceIndex(string message, char brace, int startIndex, int endIndex)
+            /// <summary>
+            /// Searches for the next brace index in the message.
+            /// </summary>
+            /// <remarks> The search skips any sequences of {{ or }}.</remarks>
+            /// <example>{{prefix{{{Argument}}}suffix}}</example>
+            /// <returns>The zero-based index position of the first occurrence of the searched brace; -1 if the searched brace was not found; -2 if the wrong brace was found.</returns>
+            private static int FindBraceIndex(string message, char searchedBrace, int startIndex, int endIndex)
             {
-                // Example: {{prefix{{{Argument}}}suffix}}.
-                int braceIndex = endIndex;
+                Debug.Assert(searchedBrace is '{' or '}');
+
+                int braceIndex = -1;
                 int scanIndex = startIndex;
-                int braceOccurrenceCount = 0;
 
                 while (scanIndex < endIndex)
                 {
-                    if (braceOccurrenceCount > 0 && message[scanIndex] != brace)
+                    char current = message[scanIndex];
+
+                    if (current is '{' or '}')
                     {
-                        if (braceOccurrenceCount % 2 == 0)
+                        char currentBrace = current;
```

Comment:
>char currentBrace = current; [](http://example.com/codeflow?start=24&length=28)

The parsing logic in general looks right, I am wondering if we can optimize a little bit? 

Can we just check here 

```C#

if (searchedBrace == '}')
{
   return current == '}' ? scanIndex : -2;
```

Then down inside the condition block of  `if (bracesCount % 2 != 0)` can be simplified like:

```C#
if (bracesCount % 2 != 0)                         
{
    return currentBrace == '{' ? scanIndex - 1 : -2;
}

```

I wrote this quickly so we may need to do a good validation if it has any defect.

Reply:
Hey @tarekgh,
So if I got it right, the proposal is:

```csharp
private static int FindBraceIndex(string message, char searchedBrace, int startIndex, int endIndex)
{
    Debug.Assert(searchedBrace is '{' or '}');

    int braceIndex = -1;
    int scanIndex = startIndex;

    while (scanIndex < endIndex)
    {
        char current = message[scanIndex];

        if (current is '{' or '}')
        {
            if (searchedBrace == '}')
            {
                return current == '}' ? scanIndex : -2;
            }

            char currentBrace = current;
            int scanIndexBeforeSkip = scanIndex;
            while (current == currentBrace && ++scanIndex < endIndex)
            {
                current = message[scanIndex];
            }

            int bracesCount = scanIndex - scanIndexBeforeSkip;
            if (bracesCount % 2 != 0)
            {
                return currentBrace == '{' ? scanIndex - 1 : -2;
            }
        }
        else
        {
            scanIndex++;
        }
    }

    return braceIndex;
}
```

If so, I added it to the solution, it passed all current tests :)

On the other hand, I think I found a defect on this new method version, that is actually "fixed" by how (the order) `ExtractTemplates` calls it.

Take for example the string template: `"}}prefix"`.

Because `ExtractTemplates` first calls `FindBraceIndex` looking for an `{`, it skips the first `}}` and because it never finds `{`, it returns `-1`, causing the method to accept it as a valid template with no arguments - which is the correct behavior here.

But if we *first* had called `FindBraceIndex` looking for an `}`, this method would return the index `0` - which is incorrect, since it should return `-1`, because it should skip the `}}`.

> the method I previously committed returns `-1`, in this case.

So, this version is indeed optimized, because it takes advantage of the previous call issue by `ExtractTemplates` - but my question is: is this temporal coupling between the method calls ok? Or the method `FindBraceIndex` should behave well "isolated"?

I'd appreciate your input here 👍 Thanks!

Reply:
Good catch regarding having a string with `}}` and not have any `{`. I think the optimization I proposed will not be solid then. I suggest keep your original version and add more test cases like `}}Something`, `something }}`, `something }}}`, `}`, and `}}`

Reply:
All right. I just added a couple of missing tests to cover the cases you mentioned.